### PR TITLE
[SPARK-48744][Core] Log entry should be constructed only once

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

In the current implementation:

```
class LogEntry(messageWithContext: => MessageWithContext) {

  def message: String = messageWithContext.message

  def context: java.util.HashMap[String, String] = messageWithContext.context
}

def logInfo(entry: LogEntry): Unit = {
  if (log.isInfoEnabled) {
    withLogContext(entry.context) {
      log.info(entry.message)
    }
  }
}
```

The field `messageWithContext` is constructed twice, one from `entry.context` and another one from `entry.message`.

This PR is to improve this and ensure a log entry is constructed only once.

### Why are the changes needed?

Improve the performance of logging

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

New UT
### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @panbingkun

**Review by HyukjinKwon — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala [hunk 1]
```diff
@@ -228,6 +228,37 @@ trait LoggingSuiteBase
           verifyMsgWithConcat(level, logOutput)
       }
   }
+
+  test("LogEntry should construct MessageWithContext only once") {
```
Comment (panbingkun) [inline common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:232]: Perhaps we should put the above testing logic into `MDCSuite` or a new UT (eg: `MessageWithContextSuite`), as it seems unrelated to the classification of `structured` and `pattern` logs, so it will be executed twice in GA?

**Comment by panbingkun** [comment]
This is really a great improvement!

**Review by yaooqinn — APPROVED** [review]

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala (hunk 1 above)
Comment (gengliangwang) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:232]: A test takes 200ms, running twice is ok. Also, we need to ensure the log entry is constructed only once when structured logging is disabled, too.

**Comment by gengliangwang** [comment]
Merging to master
